### PR TITLE
TestCase Automation : Backup and restore of namespaces with multiple provisioners using csi (vsc mapping) and kdmp

### PR DESCRIPTION
**Pull request by sn-px**
**What this PR does / why we need it**:
Automation of Test case MultipleProvisionerCsiKdmpBackupAndRestore

**Which issue(s) this PR fixes** (optional)
Closes #https://purestorage.atlassian.net/browse/PB-6278

**Special notes for your reviewer**:
Latest logs : https://aetos.pwx.purestorage.com/resultSet/testSetID/599236 
https://aetos.pwx.purestorage.com/resultSet/testSetID/599300

### CODE REVIEW

File: tests/backup/backup_multiple_provisioner_test.go [hunk 1]
```diff
@@ -222,3 +223,411 @@ var _ = Describe("{MultipleProvisionerCsiSnapshotDeleteBackupAndRestore}", Label
...
+				log.InfoD("Restoring the backup taken on singe namespace with multiple application deployed with different provisioner")
+				ctx, err := backup.GetAdminCtxFromSecret()
+				log.FailOnError(err, "Fetching px-central-admin ctx")
+				namespaceMappingMultiApp := make(map[string]string)
+				for _, appCtx := range scheduledAppContextsForMultipleAppSinleNs {
+					namespaceMappingMultiApp[appCtx.ScheduleOptions.Namespace] = appCtx.ScheduleOptions.Namespace + "-mul-app-snigle-ns"
+				}
+				restoreName := fmt.Sprintf("%s-%s-%s", "test-restore", "multi-app-single-ns", RandomString(randomStringLength))
+				log.InfoD("Restoring namespaces from the [%s] backup", multipleNsSchBackupName)
+				err = CreateRestoreWithValidation(ctx, restoreName, multipleNsSchBackupName, namespaceMappingMultiApp, make(map[string]string), DestinationClusterName, BackupOrgID, scheduledAppContextsForMultipleAppSinleNs)
```
Comment (ak-px) [inline tests/backup/backup_multiple_provisioner_test.go:537]: Restore verification is missing, Same in below restores as well.

File: tests/backup/backup_multiple_provisioner_test.go [hunk 2]
```diff
@@ -222,3 +223,411 @@ var _ = Describe("{MultipleProvisionerCsiSnapshotDeleteBackupAndRestore}", Label
...
+			forceKdmpSchBackupName, err = CreateScheduleBackupWithValidationWithVscMapping(ctx, kdmpScheduleName, SourceClusterName, backupLocationName, backupLocationUID, scheduledAppContextsForDefaultVscBackup, make(map[string]string), BackupOrgID, preRuleNameDefaultVsc, preRuleUidDefaultVsc, postRuleNameDefaultVsc, postRuleUidDefaultVsc, schedulePolicyName, schedulePolicyUID, provisionerNonDefaultSnapshotClassMap, true)
+			scheduleList = append(scheduleList, kdmpScheduleName)
+			dash.VerifyFatal(err, nil, fmt.Sprintf("Verifying creation of scheduled backup with schedule name [%s] for backup location %s", nonDefaultVscSchBackupName, backupLocationName))
+			err = IsFullBackup(forceKdmpSchBackupName, BackupOrgID, ctx)
+			dash.VerifyFatal(err, nil, fmt.Sprintf("Fetching the name of the next schedule backup for schedule: [%s] for backup location %s", nonDefaultVscSchBackupName, backupLocationName))
+			scheduleUid, err := Inst().Backup.GetBackupScheduleUID(ctx, kdmpScheduleName, BackupOrgID)
+			err = DeleteScheduleWithUIDAndWait(kdmpScheduleName, scheduleUid, SourceClusterName, srcClusterUid, BackupOrgID, ctx)
```
Comment (ak-px) [inline tests/backup/backup_multiple_provisioner_test.go:502]: Verify schedule deletion

File: tests/backup/backup_multiple_provisioner_test.go [hunk 3]
```diff
@@ -222,3 +223,411 @@ var _ = Describe("{MultipleProvisionerCsiSnapshotDeleteBackupAndRestore}", Label
...
+				nonDefaultVscSchBackupName, err = CreateScheduleBackupWithValidationWithVscMapping(ctx, nonDefaultProvisionerScheduleName, SourceClusterName, backupLocationName, backupLocationUID, scheduledAppContextsForDefaultVscBackup, make(map[string]string), BackupOrgID, preRuleNameCustomVsc, preRuleUidCustomVsc, postRuleNameCustomVsc, postRuleUidCustomVsc, schedulePolicyName, schedulePolicyUID, provisionerSnapshotClassMap, false)
+				scheduleList = append(scheduleList, nonDefaultProvisionerScheduleName)
+				dash.VerifyFatal(err, nil, fmt.Sprintf("Verifying creation of scheduled backup with schedule name [%s] for backup location %s", nonDefaultVscSchBackupName, backupLocationName))
+				err = IsFullBackup(nonDefaultVscSchBackupName, BackupOrgID, ctx)
+				dash.VerifyFatal(err, nil, fmt.Sprintf("Verifying if the first schedule backup [%s] for backup location %s is a full backup", nonDefaultVscSchBackupName, backupLocationName))
+				scheduleUid, err = Inst().Backup.GetBackupScheduleUID(ctx, nonDefaultProvisionerScheduleName, BackupOrgID)
+				err = DeleteScheduleWithUIDAndWait(nonDefaultProvisionerScheduleName, scheduleUid, SourceClusterName, srcClusterUid, BackupOrgID, ctx)
```
Comment (ak-px) [inline tests/backup/backup_multiple_provisioner_test.go:485]: schedule deletion verification

File: tests/backup/backup_multiple_provisioner_test.go [hunk 4]
```diff
@@ -222,3 +223,411 @@ var _ = Describe("{MultipleProvisionerCsiSnapshotDeleteBackupAndRestore}", Label
...
+				defaultSchBackupName, err = CreateScheduleBackupWithValidationWithVscMapping(ctx, defaultProvisionerScheduleName, SourceClusterName, backupLocationName, backupLocationUID, scheduledAppContextsForDefaultVscBackup, make(map[string]string), BackupOrgID, preRuleNameDefaultVsc, preRuleUidDefaultVsc, postRuleNameDefaultVsc, postRuleUidDefaultVsc, schedulePolicyName, schedulePolicyUID, provisionerSelectDefaultVolumeSnapshotClass, false)
+				scheduleList = append(scheduleList, defaultSchBackupName)
+
+				dash.VerifyFatal(err, nil, fmt.Sprintf("Verifying creation of scheduled backup with schedule name [%s] for backup location %s", defaultSchBackupName, backupLocationName))
+				err = IsFullBackup(defaultSchBackupName, BackupOrgID, ctx)
+				dash.VerifyFatal(err, nil, fmt.Sprintf("Verifying if the first schedule backup [%s] for backup location %s is a full backup", defaultSchBackupName, backupLocationName))
+				scheduleUid, err = Inst().Backup.GetBackupScheduleUID(ctx, defaultProvisionerScheduleName, BackupOrgID)
+				err = DeleteScheduleWithUIDAndWait(defaultProvisionerScheduleName, scheduleUid, SourceClusterName, srcClusterUid, BackupOrgID, ctx)
```
Comment (ak-px) [inline tests/backup/backup_multiple_provisioner_test.go:468]: same here

File: tests/backup/backup_multiple_provisioner_test.go [hunk 5]
```diff
@@ -222,3 +223,411 @@ var _ = Describe("{MultipleProvisionerCsiSnapshotDeleteBackupAndRestore}", Label
...
+				scheduleList = append(scheduleList, multipleNsSchBackupName)
+				dash.VerifyFatal(err, nil, fmt.Sprintf("Verifying creation of scheduled backup with schedule name [%s] for backup location %s", multipleNsSchBackupName, backupLocationName))
+				err = IsFullBackup(multipleNsSchBackupName, BackupOrgID, ctx)
+				dash.VerifyFatal(err, nil, fmt.Sprintf("Verifying if the first schedule backup [%s] for backup location %s is a full backup", multipleNsSchBackupName, backupLocationName))
+				scheduleUid, err = Inst().Backup.GetBackupScheduleUID(ctx, multipleProvisionerSameNsScheduleName, BackupOrgID)
+				err = DeleteScheduleWithUIDAndWait(multipleProvisionerSameNsScheduleName, scheduleUid, SourceClusterName, srcClusterUid, BackupOrgID, ctx)
```
Comment (ak-px) [inline tests/backup/backup_multiple_provisioner_test.go:445]: Same here

File: tests/backup/backup_multiple_provisioner_test.go [hunk 6]
```diff
@@ -222,3 +223,411 @@ var _ = Describe("{MultipleProvisionerCsiSnapshotDeleteBackupAndRestore}", Label
 		CleanupCloudSettingsAndClusters(backupLocationMap, cloudCredName, cloudCredUID, ctx)
 	})
 })
+
+var _ = Describe("{MultipleProvisionerCsiKdmpBackupAndRestore}", func() {
```
Comment (ak-px) [inline tests/backup/backup_multiple_provisioner_test.go:227]: Test doc string need to be added, Also shall we include the labels in this PR.

File: tests/backup/backup_multiple_provisioner_test.go [hunk 7]
```diff
@@ -222,3 +223,411 @@ var _ = Describe("{MultipleProvisionerCsiSnapshotDeleteBackupAndRestore}", Label
...
+
+		ctx, err := backup.GetAdminCtxFromSecret()
+		log.FailOnError(err, "Fetching px-central-admin ctx")
+		opts := make(map[string]bool)
+		opts[SkipClusterScopedObjects] = true
+
+		log.InfoD("Deleting the deployed apps after the testcase")
+		DestroyApps(allAppContext, opts)
+
+		log.InfoD("switching to default context")
```
Comment (ak-px) [inline tests/backup/backup_multiple_provisioner_test.go:627]: I think intention of deleting the destination cluster apps is not met here.

File: tests/backup/backup_multiple_provisioner_test.go (hunk 1 above)
Comment (sn-px) [reply tests/backup/backup_multiple_provisioner_test.go:534]: done

File: tests/backup/backup_multiple_provisioner_test.go [hunk 8]
```diff
@@ -222,3 +222,403 @@ var _ = Describe("{MultipleProvisionerCsiSnapshotDeleteBackupAndRestore}", func(
...
+		postRuleNameCustomVsc                     string
+		preRuleUidCustomVsc                       string
+		postRuleUidCustomVsc                      string
+		clusterProviderName                       = GetClusterProvider()
+		provisionerDefaultSnapshotClassMap        = GetProvisionerDefaultSnapshotMap(clusterProviderName)
+		provisionerSnapshotClassMap               = GetProvisionerSnapshotClassesMap(clusterProviderName)
+	)
+
+	JustBeforeEach(func() {
+		if GetClusterProvider() != "ibm" || GetClusterProvider() != "openshift" {
```
Comment (kshithijiyer-px) [inline tests/backup/backup_multiple_provisioner_test.go:276]: Can we create a list and check `Contains()`?

File: tests/backup/backup_multiple_provisioner_test.go (hunk 3 above)
Comment (sn-px) [reply tests/backup/backup_multiple_provisioner_test.go:484]: done

File: tests/backup/backup_multiple_provisioner_test.go (hunk 5 above)
Comment (sn-px) [reply tests/backup/backup_multiple_provisioner_test.go:442]: done

File: tests/backup/backup_multiple_provisioner_test.go (hunk 4 above)
Comment (sn-px) [reply tests/backup/backup_multiple_provisioner_test.go:466]: done

File: tests/backup/backup_multiple_provisioner_test.go (hunk 2 above)
Comment (sn-px) [reply tests/backup/backup_multiple_provisioner_test.go:502]: done

File: tests/backup/backup_multiple_provisioner_test.go [hunk 9]
```diff
@@ -222,3 +222,407 @@ var _ = Describe("{MultipleProvisionerCsiSnapshotDeleteBackupAndRestore}", func(
 		CleanupCloudSettingsAndClusters(backupLocationMap, cloudCredName, cloudCredUID, ctx)
 	})
 })
+
+var _ = Describe("{MultipleProvisionerCsiKdmpBackupAndRestore}", func() {
```
Comment (kshithijiyer-px) [inline tests/backup/backup_multiple_provisioner_test.go:226]: Can we go with Labels here?

File: tests/backup/backup_multiple_provisioner_test.go [hunk 10]
```diff
@@ -222,3 +222,407 @@ var _ = Describe("{MultipleProvisionerCsiSnapshotDeleteBackupAndRestore}", func(
...
+			// This test is meant to run only on IBM later will enable on other configs
+			log.Infof("Skipping the test.This test is currently configured to run on IBM and openshift and environments. Future iterations will enable on other configs")
+			Skip("Skipping the test.This test is currently configured to run on IBM and openshift environments. Future iterations will enable on other configs")
+		}
+		StartPxBackupTorpedoTest("MultipleProvisionerCsiKdmpBackupAndRestore", "Backup and restore of namespaces with multiple provisioners using csi with volume class mapping and force kdmp", nil, 296724, Sn, Q4FY24)
+		backupLocationMap = make(map[string]string)
+		labelSelectors = make(map[string]string)
+		providers = GetBackupProviders()
+		// Deploy multiple application in a single namespace using different provisioner
+		taskName := fmt.Sprintf("%s-%s", TaskNamePrefix, RandomString(randomStringLength))
```
Comment (kshithijiyer-px) [inline tests/backup/backup_multiple_provisioner_test.go:286]: Can we use time as well to be sure that the string doesn't fail if ran again?

File: tests/backup/backup_multiple_provisioner_test.go [hunk 11]
```diff
@@ -222,3 +222,407 @@ var _ = Describe("{MultipleProvisionerCsiSnapshotDeleteBackupAndRestore}", func(
...
+		})
+		Step("Registering cluster for backup", func() {
+			log.InfoD("Registering cluster for backup")
+			ctx, err := backup.GetAdminCtxFromSecret()
+			log.FailOnError(err, "Fetching px-central-admin ctx")
+
+			err = CreateApplicationClusters(BackupOrgID, "", "", ctx)
+			dash.VerifyFatal(err, nil, "Creating source and destination cluster")
+
+			clusterStatus, err := Inst().Backup.GetClusterStatus(BackupOrgID, SourceClusterName, ctx)
```
Comment (kshithijiyer-px) [inline tests/backup/backup_multiple_provisioner_test.go:354]: Can we move it into a loop?

File: tests/backup/backup_multiple_provisioner_test.go [hunk 12]
```diff
@@ -222,3 +222,407 @@ var _ = Describe("{MultipleProvisionerCsiSnapshotDeleteBackupAndRestore}", func(
...
+				restoreName := fmt.Sprintf("%s-%s-%s", "test-restore", "multi-app-single-ns", RandomString(randomStringLength))
+				log.InfoD("Restoring namespaces from the [%s] backup", multipleNsSchBackupName)
+				err = CreateRestoreWithValidation(ctx, restoreName, multipleNsSchBackupName, namespaceMappingMultiApp, make(map[string]string), DestinationClusterName, BackupOrgID, scheduledAppContextsForMultipleAppSinleNs)
+			}
+		})
+		Step("Restoring the backup taken on multiple namespace with different provisioner", func() {
+			if defaultSchBackupName != "" {
+				log.InfoD("Restoring the backup taken on multiple namespace with different provisioner")
+				ctx, err := backup.GetAdminCtxFromSecret()
+				log.FailOnError(err, "Fetching px-central-admin ctx")
```
Comment (kshithijiyer-px) [inline tests/backup/backup_multiple_provisioner_test.go:545]: Same here let's reduce repeated  and move the code into a func ?

File: tests/backup/backup_multiple_provisioner_test.go [hunk 13]
```diff
@@ -222,3 +222,407 @@ var _ = Describe("{MultipleProvisionerCsiSnapshotDeleteBackupAndRestore}", func(
...
+				log.InfoD("Skipping this step as provisioner with default volume snapshot class is not found")
+			}
+		})
+		Step(fmt.Sprintf("Creating schedule backup for multiple provisioner with non default volume snapshot class"), func() {
+			log.InfoD("Creating schedule backup for multiple provisioner with non default volume snapshot class")
+			ctx, err := backup.GetAdminCtxFromSecret()
+			log.FailOnError(err, "Fetching px-central-admin ctx")
+			if len(provisionerSnapshotClassMap) > 0 {
+				nonDefaultProvisionerScheduleName := fmt.Sprintf("default-provisioner-schedule-%v", RandomString(randomStringLength))
+				nonDefaultVscSchBackupName, err = CreateScheduleBackupWithValidationWithVscMapping(ctx, nonDefaultProvisionerScheduleName, SourceClusterName, backupLocationName, backupLocationUID, scheduledAppContextsForDefaultVscBackup, make(map[string]string), BackupOrgID, preRuleNameCustomVsc, preRuleUidCustomVsc, postRuleNameCustomVsc, postRuleUidCustomVsc, schedulePolicyName, schedulePolicyUID, provisionerSnapshotClassMap, false)
```
[the remark this answers is not in the document]
Comment (sn-px) [reply tests/backup/backup_multiple_provisioner_test.go:478]: It has multiple variations and complexity and to merge it into a loop

File: tests/backup/backup_multiple_provisioner_test.go (hunk 12 above)
Comment (sn-px) [reply tests/backup/backup_multiple_provisioner_test.go:545]: It has multiple variations and complexity and to merge it into a loop

File: tests/backup/backup_multiple_provisioner_test.go (hunk 11 above)
Comment (sn-px) [reply tests/backup/backup_multiple_provisioner_test.go:354]: It has multiple variations and complexity and to merge it into a loop

File: tests/backup/backup_multiple_provisioner_test.go [hunk 14]
```diff
@@ -222,3 +222,388 @@ var _ = Describe("{MultipleProvisionerCsiSnapshotDeleteBackupAndRestore}", func(
...
+		provisionerSnapshotClassMap               = GetProvisionerSnapshotClassesMap(clusterProviderName)
+	)
+
+	JustBeforeEach(func() {
+		if GetClusterProvider() != "ibm" || GetClusterProvider() != "openshift" {
+			// This test is meant to run only on IBM later will enable on other configs
+			log.Infof("Skipping the test.This test is currently configured to run on IBM and openshift and environments. Future iterations will enable on other configs")
+			Skip("Skipping the test.This test is currently configured to run on IBM and openshift environments. Future iterations will enable on other configs")
+		}
+		StartPxBackupTorpedoTest("MultipleProvisionerCsiKdmpBackupAndRestore", "Backup and restore of namespaces with multiple provisioners using csi with volume class mapping and force kdmp", nil, 296724, Sn, Q4FY24)
```
Comment (kphalgun-px) [inline tests/backup/backup_multiple_provisioner_test.go:281]: Q1FY25?

Comment (sn-px) [reply tests/backup/backup_multiple_provisioner_test.go]: done

File: tests/backup/backup_multiple_provisioner_test.go (hunk 9 above)
Comment (sn-px) [reply tests/backup/backup_multiple_provisioner_test.go:226]: done

File: tests/backup/backup_multiple_provisioner_test.go (hunk 8 above)
Comment (sn-px) [reply tests/backup/backup_multiple_provisioner_test.go]: done

File: tests/backup/backup_multiple_provisioner_test.go (hunk 10 above)
Comment (sn-px) [reply tests/backup/backup_multiple_provisioner_test.go:288]: done

File: tests/backup/backup_multiple_provisioner_test.go (hunk 11 above)
Comment (kshithijiyer-px) [reply tests/backup/backup_multiple_provisioner_test.go:355]: Ok

File: tests/backup/backup_multiple_provisioner_test.go (hunk 13 above)
[the remark this answers is not in the document]
Comment (kshithijiyer-px) [reply tests/backup/backup_multiple_provisioner_test.go:477]: ok

File: tests/backup/backup_multiple_provisioner_test.go [hunk 15]
```diff
@@ -222,3 +222,411 @@ var _ = Describe("{MultipleProvisionerCsiSnapshotDeleteBackupAndRestore}", Label
...
+				nonDefaultVscSchBackupName, err = CreateScheduleBackupWithValidationWithVscMapping(ctx, nonDefaultProvisionerScheduleName, SourceClusterName, backupLocationName, backupLocationUID, scheduledAppContextsForDefaultVscBackup, make(map[string]string), BackupOrgID, preRuleNameCustomVsc, preRuleUidCustomVsc, postRuleNameCustomVsc, postRuleUidCustomVsc, schedulePolicyName, schedulePolicyUID, provisionerSnapshotClassMap, false)
+				scheduleList = append(scheduleList, nonDefaultProvisionerScheduleName)
+				dash.VerifyFatal(err, nil, fmt.Sprintf("Verifying creation of scheduled backup with schedule name [%s] for backup location %s", nonDefaultVscSchBackupName, backupLocationName))
+				err = IsFullBackup(nonDefaultVscSchBackupName, BackupOrgID, ctx)
+				dash.VerifyFatal(err, nil, fmt.Sprintf("Verifying if the first schedule backup [%s] for backup location %s is a full backup", nonDefaultVscSchBackupName, backupLocationName))
+				err = SuspendBackupSchedule(nonDefaultProvisionerScheduleName, schedulePolicyName, BackupOrgID, ctx)
+				dash.VerifyFatal(err, nil, fmt.Sprintf("Verification of deleting backup scheudle [%s]", nonDefaultProvisionerScheduleName))
```
[the remark this answers is not in the document]
Comment (sn-px) [reply tests/backup/backup_multiple_provisioner_test.go]: done

**Review by kphalgun-px — APPROVED** [review]
LGTM

File: tests/backup/backup_multiple_provisioner_test.go [hunk 16]
```diff
@@ -222,3 +222,410 @@ var _ = Describe("{MultipleProvisionerCsiSnapshotDeleteBackupAndRestore}", Label
...
+			log.FailOnError(err, "failed to SetClusterContext to default cluster")
+		}()
+
+		ctx, err := backup.GetAdminCtxFromSecret()
+		for _, scheduleName := range scheduleNames {
+			err = DeleteSchedule(scheduleName, SourceClusterName, BackupOrgID, ctx)
+			dash.VerifySafely(err, nil, fmt.Sprintf("Verification of deleting backup schedule - %s", scheduleName))
+		}
+
+		log.FailOnError(err, "Fetching px-central-admin ctx")
```
Comment (ak-px) [inline tests/backup/backup_multiple_provisioner_test.go:622]: need to move the verification to line 617

Comment (sn-px) [reply tests/backup/backup_multiple_provisioner_test.go]: done

**Review by kshithijiyer-px — APPROVED** [review]
lgtm

**Pull request merged**